File: apps/api/app/schemas/record.py
"""Request/response shapes for records, query, and aggregate (plan sections
9.1, 10.4, 21.2; docs/API.md §5).

`CreateRecordRequest.data`/`UpdateRecordRequest.data` are deliberately plain
`dict[str, Any]` here — the *real* validation is the page-specific model
`app/schemas/dynamic.py` builds from that page's live columns, which this
generic request shape can't know about. `RecordOut.data` is likewise passed
through as stored: every value already lives in `records.data` in wire
format (money as a string, dates as ISO text — plan section 9.1), so no
re-serialization happens on the way out.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.db.base import RecordStatus
from app.schemas.common import QueryFilter, SortSpec


class CreateRecordRequest(BaseModel):
    occurred_at: datetime
    client_uuid: uuid.UUID | None = None
    store_id: uuid.UUID | None = None
    data: dict[str, Any] = Field(default_factory=dict)


class UpdateRecordRequest(BaseModel):
    #: Optimistic-locking version, as an alternative to the `If-Match` header
    #: (docs/API.md §1.2) — either is accepted; the router resolves which one
    #: was actually supplied.
    version: int | None = None
    occurred_at: datetime | None = None
    store_id: uuid.UUID | None = None
    data: dict[str, Any] = Field(default_factory=dict)


class DeleteRecordRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=1000)


class SetProtectedFieldRequest(BaseModel):
    """docs/API.md §5 / plan section 11.4 — the only body a protected
    column ever accepts: which column, its new value, and the version
    being changed from (optimistic locking, same as any other update).
    `column_key` matches docs/API.md's documented field name exactly (a P4
    fix — the field was originally shipped as `column`)."""

    column_key: str
    value: str
    version: int | None = None


class RecordOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    page_id: uuid.UUID
    store_id: uuid.UUID | None
    occurred_at: datetime
    business_date: date
    status: RecordStatus
    data: dict[str, Any]
    needs_review: bool
    version: int
    created_by: uuid.UUID
    updated_by: uuid.UUID | None
    #: Set only on a reversal's replacement record (P4 §8) — always `None`
    #: on a native/system-table row, and on every ordinary record.
    reverses_id: uuid.UUID | None = None


class ReverseRecordRequest(BaseModel):
    """`POST /records/{id}/reverse` (plan section 11.5, P4 §8) — `kind=LEDGER`
    pages only. `version` is the *original* record's optimistic-locking
    version. Omitting `data` means "void this record, nothing replaces it";
    supplying it creates a new `ACTIVE` record linked via `reverses_id`,
    validated/formula-computed exactly like an ordinary create
    (`record_service.create_record`'s own pipeline, not a parallel one)."""

    version: int
    data: dict[str, Any] | None = None
    occurred_at: datetime | None = None
    store_id: uuid.UUID | None = None


class ReverseRecordResponse(BaseModel):
    original: RecordOut
    replacement: RecordOut | None = None


class RunningBalanceEntry(BaseModel):
    record_id: uuid.UUID
    business_date: date
    #: The balance column's own value on this row (wire format, a money
    #: string) — alongside the cumulative total, so a client can render both
    #: without a second round trip.
    amount: str
    balance: str


class RunningBalanceResponse(BaseModel):
    entries: list[RunningBalanceEntry] = Field(default_factory=list)


class ExportRequest(BaseModel):
    """`POST /pages/{id}/export` (plan section 13.2 / P3.5 Part 2). Deliberately
    a subset of `QueryRequest`'s fields: an export has no `cursor`/`limit`
    (every matching row goes out) and no client-chosen `sort` (rows are
    always written in `business_date` order)."""

    filters: list[QueryFilter] = Field(default_factory=list)
    search: str | None = None


class QueryRequest(BaseModel):
    filters: list[QueryFilter] = Field(default_factory=list)
    search: str | None = None
    sort: list[SortSpec] = Field(default_factory=list)
    cursor: str | None = None
    limit: int = Field(default=50, ge=1, le=500)


#: docs/API.md §5: sum, avg, count, min, max.
AggregateMetric = Literal["sum", "avg", "count", "min", "max"]
#: Server-resolved relative windows over `business_date` — kept small and
#: explicit rather than accepting arbitrary date math from the client.
AggregatePeriod = Literal["current_month", "last_month", "current_year", "all_time"]


class AggregateRequest(BaseModel):
    metric: AggregateMetric
    #: Required for every metric except `count`, which may count whole rows.
    column: str | None = None
    group_by: str | None = None
    period: AggregatePeriod = "all_time"
    filters: list[QueryFilter] = Field(default_factory=list)


class AggregateGroup(BaseModel):
    key: str
    value: str
    count: int


class AggregatePeriodRange(BaseModel):
    from_: date | None = Field(default=None, alias="from")
    to: date | None = None

    model_config = ConfigDict(populate_by_name=True)


class AggregateResponse(BaseModel):
    value: str
    record_count: int
    period: AggregatePeriodRange
    groups: list[AggregateGroup] = Field(default_factory=list)


class ColumnValuesResponse(BaseModel):
    values: list[str]

File: apps/mobile/lib/features/pages/domain/record.dart
import 'column.dart';

/// One row of an Owner-defined page (plan section 8.3).
///
/// Named `PageRecord`, not `Record` — `Record` is a built-in Dart 3 type.
///
/// Mirrors `RecordOut` in `app/schemas/record.py`. `data` holds the Owner's
/// own columns keyed by column `key`, already in **wire format**: money and
/// numbers as strings, dates as ISO text, booleans as real booleans (plan
/// section 9.1). Nothing here converts them — `AmountText` and the field
/// renderers do that at the edge where they are displayed or edited.
class PageRecord {
  const PageRecord({
    required this.id,
    required this.companyId,
    required this.pageId,
    required this.occurredAt,
    required this.businessDate,
    required this.status,
    required this.data,
    required this.needsReview,
    required this.version,
    required this.createdBy,
    this.storeId,
    this.updatedBy,
  });

  factory PageRecord.fromJson(Map<String, dynamic> json) {
    return PageRecord(
      id: json['id'] as String,
      companyId: json['company_id'] as String,
      pageId: json['page_id'] as String,
      storeId: json['store_id'] as String?,
      occurredAt: json['occurred_at'] as String,
      businessDate: json['business_date'] as String,
      status: RecordStatus.fromWire(json['status'] as String),
      data: (json['data'] as Map<String, dynamic>? ?? const {}),
      needsReview: json['needs_review'] as bool,
      version: json['version'] as int,
      createdBy: json['created_by'] as String,
      updatedBy: json['updated_by'] as String?,
    );
  }

  final String id;
  final String companyId;
  final String pageId;
  final String? storeId;

  /// When the business event happened — ISO 8601 with offset.
  final String occurredAt;

  /// The accounting day, derived server-side (plan section 9.2). Read-only.
  final String businessDate;

  final RecordStatus status;

  /// The Owner's columns, keyed by column `key`, in wire format.
  final Map<String, dynamic> data;

  /// Set by a WARNING validation rule (plan section 11.3). Rules themselves
  /// arrive with P4; the flag is displayed here when the server sets it.
  final bool needsReview;

  /// Optimistic-locking version — sent back as `If-Match` on edit.
  final int version;

  final String createdBy;
  final String? updatedBy;

  /// The stored value for a column, or null.
  Object? valueFor(PageColumn column) => data[column.key];
}

/// The platform record lifecycle (plan section 4.4) — distinct from any
/// Owner-defined status column, which is why `cheques` names its own one
/// `cheque_status` (docs/API.md section 1.7).
enum RecordStatus {
  active,
  reversed,
  void_;

  static RecordStatus fromWire(String value) => switch (value) {
    'ACTIVE' => RecordStatus.active,
    'REVERSED' => RecordStatus.reversed,
    'VOID' => RecordStatus.void_,
    _ => throw ArgumentError('Unknown record status: $value'),
  };

  String get label => switch (this) {
    RecordStatus.active => 'Active',
    RecordStatus.reversed => 'Reversed',
    RecordStatus.void_ => 'Void',
  };

  String get wire => switch (this) {
    RecordStatus.active => 'ACTIVE',
    RecordStatus.reversed => 'REVERSED',
    RecordStatus.void_ => 'VOID',
  };
}

// --------------------------------------------------------------------------
// Query language (plan section 10.4)
// --------------------------------------------------------------------------

/// The structured filter operators the server accepts. Column keys are
/// validated against `page_columns` server-side and mapped to safe
/// expressions — never interpolated into SQL (plan section 3.8).
enum FilterOp {
  eq,
  neq,
  gt,
  gte,
  lt,
  lte,
  between,
  inList,
  contains,
  isNull;

  String get wire => switch (this) {
    FilterOp.eq => 'eq',
    FilterOp.neq => 'neq',
    FilterOp.gt => 'gt',
    FilterOp.gte => 'gte',
    FilterOp.lt => 'lt',
    FilterOp.lte => 'lte',
    FilterOp.between => 'between',
    FilterOp.inList => 'in',
    FilterOp.contains => 'contains',
    FilterOp.isNull => 'is_null',
  };

  String get label => switch (this) {
    FilterOp.eq => 'is',
    FilterOp.neq => 'is not',
    FilterOp.gt => 'more than',
    FilterOp.gte => 'at least',
    FilterOp.lt => 'less than',
    FilterOp.lte => 'at most',
    FilterOp.between => 'between',
    FilterOp.inList => 'is one of',
    FilterOp.contains => 'contains',
    FilterOp.isNull => 'is empty',
  };

  /// Which operators make sense for a column type. Offering "more than" on a
  /// yes/no column is how a filter sheet becomes confusing.
  static List<FilterOp> forType(ColumnType type) => switch (type) {
    ColumnType.text || ColumnType.longText => const [
      FilterOp.eq,
      FilterOp.neq,
      FilterOp.contains,
      FilterOp.isNull,
    ],
    ColumnType.number || ColumnType.currency || ColumnType.percent => const [
      FilterOp.eq,
      FilterOp.neq,
      FilterOp.gt,
      FilterOp.gte,
      FilterOp.lt,
      FilterOp.lte,
      FilterOp.between,
      FilterOp.isNull,
    ],
    ColumnType.date || ColumnType.datetime => const [
      FilterOp.eq,
      FilterOp.gte,
      FilterOp.lte,
      FilterOp.between,
      FilterOp.isNull,
    ],
    ColumnType.boolean => const [FilterOp.eq, FilterOp.isNull],
    ColumnType.select || ColumnType.multiSelect => const [
      FilterOp.eq,
      FilterOp.neq,
      FilterOp.inList,
      FilterOp.isNull,
    ],
    ColumnType.recordRef || ColumnType.storeRef || ColumnType.userRef => const [
      FilterOp.eq,
      FilterOp.neq,
      FilterOp.isNull,
    ],
    // Not filterable in P3: FORMULA has no stored value until P4's evaluator,
    // and ATTACHMENT holds a count the attachment endpoints (P5) maintain.
    ColumnType.formula || ColumnType.attachment => const [],
  };
}

class RecordFilter {
  const RecordFilter({required this.column, required this.op, this.value});

  final String column;
  final FilterOp op;

  /// A scalar, a 2-element list for `between`, a list for `in`, or null for
  /// `is_null`. Always in wire format.
  final Object? value;

  Map<String, dynamic> toJson() => {
    'column': column,
    'op': op.wire,
    'value': value,
  };
}

class SortSpec {
  const SortSpec({required this.column, this.descending = false});

  final String column;
  final bool descending;

  Map<String, dynamic> toJson() => {
    'column': column,
    'direction': descending ? 'desc' : 'asc',
  };
}

/// The body of `POST /pages/{id}/records/query`.
class RecordQuery {
  const RecordQuery({
    this.filters = const [],
    this.sort = const [],
    this.search,
    this.cursor,
    this.limit = 50,
  });

  final List<RecordFilter> filters;
  final List<SortSpec> sort;
  final String? search;
  final String? cursor;
  final int limit;

  RecordQuery copyWith({
    List<RecordFilter>? filters,
    List<SortSpec>? sort,
    String? search,
    String? cursor,
    int? limit,
    bool clearCursor = false,
    bool clearSearch = false,
  }) {
    return RecordQuery(
      filters: filters ?? this.filters,
      sort: sort ?? this.sort,
      search: clearSearch ? null : (search ?? this.search),
      cursor: clearCursor ? null : (cursor ?? this.cursor),
      limit: limit ?? this.limit,
    );
  }

  Map<String, dynamic> toJson() => {
    'filters': filters.map((f) => f.toJson()).toList(),
    'sort': sort.map((s) => s.toJson()).toList(),
    if (search != null && search!.isNotEmpty) 'search': search,
    if (cursor != null) 'cursor': cursor,
    'limit': limit,
  };

  /// Two queries describe the same result set when everything but the cursor
  /// matches — used to know when to reset pagination rather than append.
  bool sameShapeAs(RecordQuery other) {
    if (search != other.search || limit != other.limit) return false;
    if (filters.length != other.filters.length) return false;
    if (sort.length != other.sort.length) return false;
    for (var i = 0; i < filters.length; i++) {
      final a = filters[i];
      final b = other.filters[i];
      if (a.column != b.column || a.op != b.op || '${a.value}' != '${b.value}') {
        return false;
      }
    }
    for (var i = 0; i < sort.length; i++) {
      if (sort[i].column != other.sort[i].column ||
          sort[i].descending != other.sort[i].descending) {
        return false;
      }
    }
    return true;
  }
}

/// `sum, avg, count, min, max` (plan section 10.4), executed in Postgres —
/// the client never sums a column itself.
enum AggregateMetric {
  sum,
  avg,
  count,
  min,
  max;

  String get wire => name;
}

/// The body of `POST /pages/{id}/aggregate`.
class AggregateQuery {
  const AggregateQuery({
    required this.metric,
    this.column,
    this.groupBy,
    this.period = 'all_time',
    this.filters = const [],
  });

  final AggregateMetric metric;
  final String? column;
  final String? groupBy;

  /// `current_month` · `last_month` · `current_year` · `all_time`.
  final String period;
  final List<RecordFilter> filters;

  Map<String, dynamic> toJson() => {
    'metric': metric.wire,
    if (column != null) 'column': column,
    if (groupBy != null) 'group_by': groupBy,
    'period': period,
    'filters': filters.map((f) => f.toJson()).toList(),
  };
}

class AggregateGroup {
  const AggregateGroup({required this.key, required this.value, required this.count});

  factory AggregateGroup.fromJson(Map<String, dynamic> json) => AggregateGroup(
    key: json['key'] as String,
    value: json['value'] as String,
    count: json['count'] as int,
  );

  final String key;

  /// Always a wire-format string (money stays a string all the way through).
  final String value;
  final int count;
}

class AggregateResult {
  const AggregateResult({
    required this.value,
    required this.recordCount,
    required this.groups,
  });

  factory AggregateResult.fromJson(Map<String, dynamic> json) => AggregateResult(
    value: json['value'] as String,
    recordCount: json['record_count'] as int,
    groups: (json['groups'] as List<dynamic>? ?? const [])
        .cast<Map<String, dynamic>>()
        .map(AggregateGroup.fromJson)
        .toList(),
  );

  final String value;
  final int recordCount;
  final List<AggregateGroup> groups;
}

/// The cursor-paginated envelope every list endpoint returns (docs/API.md
/// section 1.3).
class Paginated<T> {
  const Paginated({required this.items, this.nextCursor, this.hasMore = false});

  factory Paginated.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) parseItem,
  ) {
    return Paginated(
      items: (json['items'] as List<dynamic>? ?? const [])
          .cast<Map<String, dynamic>>()
          .map(parseItem)
          .toList(),
      nextCursor: json['next_cursor'] as String?,
      hasMore: json['has_more'] as bool? ?? false,
    );
  }

  final List<T> items;
  final String? nextCursor;
  final bool hasMore;
}
